### ISSUE TITLE
[onert] Automatic backend resolver (per operation type) as default runtime behavior

### ISSUE THREAD

**Issue by wateret** [Samsung/ONE#3891]
We thought the runtime would find an available backend for each operation and automatically use that backend as a default behavior of the runtime. 

However the current runtime just tries the first loaded backend that is given to `BACKENDS` or `nnfw_set_available_backends`. And if it fails compilation with that backend, it just fails even if there are more backends available that can run those operations.

I'm thinking of a scenario like this. Let's say we have backend list in the order - cpu, acl_cl and acl neon. And those supports different set of operations.

- cpu supports Add and Sub
- acl_cl supports Sub and Mul
- acl_neon supports Div

Runtime choose a backend for an operation in the order of the given list.

- Add : cpu - as cpu supports Add
- Sub : cpu - as cpu supports Sub, acl_cl also supports it but cpu has priority
- Mul : acl_cl - as cpu does not support Mul but acl_cl supports it
- Div : acl_neon - as cpu and acl_cl do not support Div but acl_neon supports it

However how can we decide which backend supports which operations? It is not that simple to know since even though for the same type of operation, some backends could support it partially. For example, some backends could support Conv2D operation without dilations.

So I would suggest this way, but it is not perfect, though.

### Try compilation and throw when it can't

- Try compilation (run IKernelGenerator)
    - If success, that backend is chosen
    - If failure, try compilation again with the next backend

With this way, as it tries compilation which could take relatively long.

---

Ideas welcome.

**Comment by wateret** [comment]
cc: @dayo09

**Comment by semun-lee** [comment]
This idea looks reasonable. Actually, when we run posenet which is one of the models for Tizen mediavision API, we needed to specify operators which should fallback to cpu backend manually. If runtime fallbacks automatically, we could run models with various backend configurations easily.

`
BACKENDS="acl_cl;cpu"  OP_BACKEND_SplitV="cpu" OP_BACKEND_Reverse="cpu" OP_BACKEND_SpaceToBatchND="cpu" OP_BACKEND_BatchToSpaceND="cpu" OP_BACKEND_ArgMax="cpu" `

**Comment by jinevening** [comment]
Since a DNN model is not likely to change (unless it is trained online), it would be possible to save a list of available backends for each operator of the model as a file and reuse it in the next run.

> Sub : cpu - as cpu supports Sub, acl_cl also supports it but cpu has priority

As a future work, it would be great if the priority is determined by actual profiling results in the target machine.

For example, if a model is given, onert can record the execution time of each operator (layer) with different backends (only for the backends available to run the operator). The recorded execution time can be saved as a file, used for backend selection afterwards. e.g., faster backend would have higher priority.

The recorded execution time may be uselful for HEScheduler as well.

**Comment by s-barannikov** [comment]
A backend could provide an interface which tells whether an operation is supported or not (by that backend).
Visitor pattern should do. The `visit` methods examines a concrete operation (dilations etc.) and returns true / false.
But that only works if all the necessary parameters are known at the compile time. If, for example, some operation cannot work with `axis` other than 1, and the `axis` is the non-constant input tensor, this won't work.

**Comment by wateret** [comment]
@s-barannikov I thought of that too, but what do you think about this?

> However how can we decide which backend supports which operations? It is not that simple to know since even though for the same type of operation, a same backend could compile it or not, depending on the parameters. For example, some backends can only support Conv2D operation without dilations.

This is why I wanted try-compilation-or-throw to check an operation is supported.

**Comment by wateret** [comment]
@jinevening 

> For example, if a model is given, onert can record the execution time of each operator (layer) with different backends (only for the backends available to run the operator). The recorded execution time can be saved as a file, used for backend selection afterwards. e.g., faster backend would have higher priority.
> 
> The recorded execution time may be uselful for HEScheduler as well.

That is pretty much what `HEScheduler` does already 😄

**Comment by s-barannikov** [comment]
> @s-barannikov I guess what I stated above is pretty much same with what you said.

Almost, with the exception that it does not throw exceptions.

> It is why I would like try-compilation-or-throw to check an operation is supported.

It has the same issue with non-constant axis and alike.

**Comment by wateret** [comment]
> > It is why I would like try-compilation-or-throw to check an operation is supported.
> 
> It has the same issue with non-constant axis and alike.

Let me clarify terms first.

- Params : Always const
- Input tensors : Tensors which could be const and non-const

So the params cannot be non-const, it is not a thing.
If we had axis(or so) as an input tensor, if we know which is const or not at compilation time, so for those backends that don't support dynamic axis, it would throw. So don't see any issues. Please correct me if I'm wrong.

**Comment by s-barannikov** [comment]
@wateret 
Sorry for confusion. I mean axis as non-constant input tensor. Some backend could support it, but only when axis is equal to 1 or 2, which can be known only at execution time. E.g. it will throw during execution rather than at configure time. I thought I saw this in some ACL kernels (most probably CL ones), but I can't find it right now.
From this point of view, either variant will not work (yours or mine). Such cases will probably be rare though.

**Comment by wateret** [comment]
@s-barannikov  Thank you for the details.

What I'm saying is a suggestion for the near future rather than as-is. Getting back to your example, the axis tensor is non-const input. For the backends that do not support dynamically changing axis, then it must throw during configuration as we know it is non-const even though we don't know its value at configuration, rather than go on and crash during execution. If the current ACL implementation does not this, we can make it by revising KernelGenerator - this is what I'm suggesting.

**Comment by s-barannikov** [comment]
I searched more carefully for the piece of code in ACL, but could not find one. Probably I'm imagining things, or saw it elsewhere.

> For the backends that do not support dynamically changing axis, then it must throw during configuration as we know it is non-const even though we don't know its value at configuration, rather than go on and crash during execution. <...> Does it make sense to you?

It does. But I would not throw exceptions, but instead return true/false as I suggested earlier. IMHO exceptions are for exceptional (unexpected) conditions, they should not be used for normal control flow. They are also _much_ slower than returning a boolean value.

**Comment by wateret** [comment]
> But I would not throw exceptions, but instead return true/false as I suggested earlier. IMHO exceptions are for exceptional (unexpected) conditions, they should not be used for normal control flow. They are also much slower than returning a boolean value.

Agreed on returning bool. I said `throw` since I just wanted to avoid having visitor accept functions with different return types or storing return value in the visitor, which creates a redundant state(yes, we have some code like this already 😅 ). Anyways returning `bool` looks better here and I'm thinking about the later one(storing return value in the visitor).

**Comment by s-barannikov** [comment]
> I just wanted to avoid having visitor accept functions with different return types or storing return value in the visitor, which creates a redundant state(yes, we have some code like this already 😅 ).

I don't mind the state, but it is easy to forget to update it in one of the `visit` methods :(

> avoid having visitor accept functions with different return types

In case I didn't get it right, `loco` implements a templated visitor, which can be used like this:

```C++
class MyVisitor1 : public Visitor<bool> {
  bool visit(...);
};
class MyVisitor2 : public Visitor<void> {
  voidvisit(...);
};
```

It also requires the `accept` method to be templated, resulting in extra instantiations (one for each return type).
I can't say that I like it (my IDE does not parse templates well, disabling some useful features 🤣 ), but may be you'll find it acceptable.

I don't know the other way except the above two. Switch does not count :D

**Comment by s-barannikov** [comment]
> it is easy to forget to update it in one of the `visit` methods :(

Unit tests could help

**Comment by wateret** [comment]
@s-barannikov What I thought was, we can templatize visitors easily. But for `accept` functions in visitees, every time we need another visitor that returns a new type, we also need to edit the visitee(For this case it is bool, but I'm talking about in general).
If a visitee is coupled to a type that a specific visitor returns, then I'm not sure it is worth it. Maybe introducing a virtual function for it would be better.

I'm not sure about this so please correct me if I'm wrong on something. 😄

**Comment by s-barannikov** [comment]
> every time we need another visitor that returns a new type, we also need to edit the visitee

The `accept` could also be templated. But
1. A templated method cannot be virtual (and thus cannot be overloaded), so one will have to do with switch instead of classical double dispatch:
```C++
class Operation {
  template <typename ResultT> ResultT accept(TemplatedVisitor<ResultT> *v)
  {
    switch (this->opcode()) {
      case OpCode::Conv2D:
        return v->visit(dynamic_cast<operation::Conv2D *>(this));
      // ...
    }
  }
};
```
This will work for both `void` and `bool`.

2. Since it is a template, it should be either put in header file, or explicit instantiations must be done for every needed type.

I don't like both points.

**Comment by s-barannikov** [comment]
> Maybe introducing a virtual function for it would be better.

Sounds interesting. Could you show what you have in mind? Function prototype would do :)

**Comment by wateret** [comment]
> The `accept` could also be templated. But
> 1. A templated method cannot be virtual (and thus cannot be overloaded), so one will have to do with switch instead of classical double dispatch:

Oh I just remembered, long ago, why I stopped trying templatization of visitors - tempate methods cannot be virtual. Thank you for letting me know we could do that way. Now I see why `loco` implemented `accept` function just like that.

> 2. Since it is a template, it should be either put in header file, or explicit instantiations must be done for every needed type.

Looks like that is something that we cannot avoid. I would choose the former, if we are templatizing visitors.

**Comment by wateret** [comment]
> > Maybe introducing a virtual function for it would be better.
> 
> Sounds interesting. Could you show what you have in mind? Function prototype would do :)

Never mind 😄 , as it would cause depedency hell since operation will be dependent to backend module. It was just to say this way might be better than when "a visitee is coupled to a type that a specific visitor returns".
